File: App/Source/Authored/Scripts/Engine/UI/EditText.cs
// Represents a Falcon::EditTextInstance in script. Most often,
// used to call SetText() or SetTextToken() to update the body
// of an EditText instance.
//
// Copyright (c) Demiurge Studios, Inc.
// 
// This source code is licensed under the MIT license.
// Full license details can be found in the LICENSE file
// in the root directory of this source tree.

using static SlimCS;

public static class FocusEvent
{
	public const string FOCUS_IN = "focusIn";
}

public sealed class EditText : DisplayObject
{
	/// <summary>
	/// Convenience - resturns the text of an EditText
	/// restricted to its EditingMaxChars value, if set.
	/// </summary>
	/// <param name="txt">The edit text to apply.</param>
	/// <returns>The limited text value.</returns>
	public string LimitedText
	{
		get
		{
			var sText = GetText();
			var iChars = EditingMaxChars ?? 0;
			if (iChars >= 0)
			{
				sText = CoreNative.StringSub(sText, 0, iChars);
			}

			return sText;
		}
	}

	public void CommitFormatting()
	{
		dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).CommitFormatting();
	}

	public bool GetAutoSizeBottom()
	{
		return dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).GetAutoSizeBottom();
	}

	public bool GetAutoSizeContents()
	{
		return dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).GetAutoSizeContents();
	}

	public bool GetAutoSizeHorizontal()
	{
		return dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).GetAutoSizeHorizontal();
	}

	public (int, int, int, int) GetCursorColor()
	{
		return dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).GetCursorColor();
	}

	public bool GetHasTextEditFocus()
	{
		return dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).GetHasTextEditFocus();
	}

	public int GetNumLines()
	{
		return dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).GetNumLines();
	}

	public string GetPlainText()
	{
		return dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).GetPlainText();
	}

	public string GetText()
	{
		return dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).GetText();
	}

	public string GetXhtmlText()
	{
		return dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).GetXhtmlText();
	}

	public bool GetVerticalCenter()
	{
		return dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).GetVerticalCenter();
	}

	public bool GetXhtmlParsing()
	{
		return dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).GetXhtmlParsing();
	}

	public (double, double, double, double) GetTextBounds()
	{
		return ((double, double, double, double))dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).GetTextBounds();
	}

	public double GetTextHeight()
	{
		var (_, t, _, b) = GetTextBounds();
		return b - t;
	}

	public double GetTextWidth()
	{
		var (l, _, r, _) = GetTextBounds();
		return r - l;
	}

	public (double, double, double, double) GetLocalTextBounds()
	{
		return ((double, double, double, double))dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).GetLocalTextBounds();
	}

	public double GetVisibleCharacters()
	{
		return dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).GetVisibleCharacters();
	}

	public (double, double, double, double) GetWorldTextBounds()
	{
		return ((double, double, double, double))dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).GetWorldTextBounds();
	}

	public void SetAutoSizeBottom(bool bAutoSizeBottom)
	{
		dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).SetAutoSizeBottom(bAutoSizeBottom);
	}

	public void SetAutoSizeContents(bool bAutoSizeContents)
	{
		dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).SetAutoSizeContents(bAutoSizeContents);
	}

	public void SetAutoSizeHorizontal(bool bAutoSizeHorizontal)
	{
		dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).SetAutoSizeHorizontal(bAutoSizeHorizontal);
	}

	public void SetCursorColor(int iR, int iG, int iB, int iA)
	{
		dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).SetCursorColor(iR, iG, iB, iA);
	}

	public void SetPlainText(string sText)
	{
		dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).SetPlainText(sText);
	}

	public void SetText(string sText)
	{
		dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).SetText(sText);
	}

	public void SetTextToken(string sTextToken)
	{
		dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).SetTextToken(sTextToken);
	}

	public void SetVerticalCenter(bool bVerticalCenter)
	{
		dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).SetVerticalCenter(bVerticalCenter);
	}

	public void SetVisibleCharacters(int chars)
	{
		dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).SetVisibleCharacters(chars);
	}

	public void SetXhtmlParsing(bool bXhtmlParsing)
	{
		dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).SetXhtmlParsing(bXhtmlParsing);
	}

	public void SetXhtmlText(string sText)
	{
		dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).SetXhtmlText(sText);
	}

	/// <summary>
	/// Optional - used if StartEditing() without complete arguments is called.
	/// </summary>
	public string EditingDescription { get; set; }

	/// <summary>
	/// Optional - used if StartEditing() without complete arguments is called.
	/// </summary>
	public int? EditingMaxChars { get; set; }

	/// <summary>
	/// Optional - used if StartEditing() without complete arguments is called.
	/// </summary>
	public string EditingRestrictedCharacters { get; set; }

	public bool StartEditing(
		MovieClip eventReceiver,
		bool bAllowNonLatinKeyboard = false)
	{
		var sDescription = EditingDescription ?? string.Empty;
		var iMaxCharacters = EditingMaxChars ?? 0;
		var sRestrict = EditingRestrictedCharacters ?? string.Empty;

		return StartEditing(eventReceiver, sDescription, iMaxCharacters, sRestrict, bAllowNonLatinKeyboard);
	}

	public bool StartEditing(
		MovieClip eventReceiver,
		string sDescription,
		int iMaxCharacters,
		string sRestrict,
		bool bAllowNonLatinKeyboard = false)
	{
		var bReturn = dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).StartEditing(
			dyncast<Native.ScriptUIMovieClipInstance>(eventReceiver.NativeInstance),
			sDescription,
			iMaxCharacters,
			sRestrict,
			bAllowNonLatinKeyboard);

		// Focus in on first start/successful start.
		if (bReturn)
		{
			DispatchEvent(this, FocusEvent.FOCUS_IN);
		}

		return bReturn;
	}

	public void StopEditing()
	{
		dyncast<Native.ScriptUIEditTextInstance>(m_udNativeInstance).StopEditing();
	}
}

File: SeoulTools/Data/Config/UnitTests/SlimCS/ScriptsDebug/Engine/UI/EditText.lua
-- Represents a Falcon::EditTextInstance in script. Most often,
-- used to call SetText() or SetTextToken() to update the body
-- of an EditText instance.
--
-- Copyright (c) Demiurge Studios, Inc.
-- 
-- This source code is licensed under the MIT license.
-- Full license details can be found in the LICENSE file
-- in the root directory of this source tree.



local FocusEvent = class_static 'FocusEvent'

FocusEvent.FOCUS_IN = 'focusIn'


local EditText = class('EditText', 'DisplayObject')

--- <summary>
--- Convenience - resturns the text of an EditText
--- restricted to its EditingMaxChars value, if set.
--- </summary>
--- <param name="txt">The edit text to apply.</param>
--- <returns>The limited text value.</returns>


function EditText:get_LimitedText()

	local sText = self:GetText()
	local iChars = self.EditingMaxChars or 0
	if iChars >= 0 then

		sText = CoreNative.StringSub(sText, 0, iChars)
	end

	return sText
end


function EditText:CommitFormatting()

	self.m_udNativeInstance:CommitFormatting()
end

function EditText:GetAutoSizeBottom()

	return self.m_udNativeInstance:GetAutoSizeBottom()
end

function EditText:GetAutoSizeContents()

	return self.m_udNativeInstance:GetAutoSizeContents()
end

function EditText:GetAutoSizeHorizontal()

	return self.m_udNativeInstance:GetAutoSizeHorizontal()
end

function EditText:GetCursorColor()

	return self.m_udNativeInstance:GetCursorColor()
end

function EditText:GetHasTextEditFocus()

	return self.m_udNativeInstance:GetHasTextEditFocus()
end

function EditText:GetNumLines()

	return self.m_udNativeInstance:GetNumLines()
end

function EditText:GetPlainText()

	return self.m_udNativeInstance:GetPlainText()
end

function EditText:GetText()

	return self.m_udNativeInstance:GetText()
end

function EditText:GetXhtmlText()

	return self.m_udNativeInstance:GetXhtmlText()
end

function EditText:GetVerticalCenter()

	return self.m_udNativeInstance:GetVerticalCenter()
end

function EditText:GetXhtmlParsing()

	return self.m_udNativeInstance:GetXhtmlParsing()
end

function EditText:GetTextBounds()

	return self.m_udNativeInstance:GetTextBounds()
end

function EditText:GetTextHeight()

	local _, t, _, b = self:GetTextBounds()
	return b - t
end

function EditText:GetTextWidth()

	local l, _, r, _ = self:GetTextBounds()
	return r - l
end

function EditText:GetLocalTextBounds()

	return self.m_udNativeInstance:GetLocalTextBounds()
end

function EditText:GetVisibleCharacters()

	return self.m_udNativeInstance:GetVisibleCharacters()
end

function EditText:GetWorldTextBounds()

	return self.m_udNativeInstance:GetWorldTextBounds()
end

function EditText:SetAutoSizeBottom(bAutoSizeBottom)

	self.m_udNativeInstance:SetAutoSizeBottom(bAutoSizeBottom)
end

function EditText:SetAutoSizeContents(bAutoSizeContents)

	self.m_udNativeInstance:SetAutoSizeContents(bAutoSizeContents)
end

function EditText:SetAutoSizeHorizontal(bAutoSizeHorizontal)

	self.m_udNativeInstance:SetAutoSizeHorizontal(bAutoSizeHorizontal)
end

function EditText:SetCursorColor(iR, iG, iB, iA)

	self.m_udNativeInstance:SetCursorColor(iR, iG, iB, iA)
end

function EditText:SetPlainText(sText)

	self.m_udNativeInstance:SetPlainText(sText)
end

function EditText:SetText(sText)

	self.m_udNativeInstance:SetText(sText)
end

function EditText:SetTextToken(sTextToken)

	self.m_udNativeInstance:SetTextToken(sTextToken)
end

function EditText:SetVerticalCenter(bVerticalCenter)

	self.m_udNativeInstance:SetVerticalCenter(bVerticalCenter)
end

function EditText:SetVisibleCharacters(chars)

	self.m_udNativeInstance:SetVisibleCharacters(chars)
end

function EditText:SetXhtmlParsing(bXhtmlParsing)

	self.m_udNativeInstance:SetXhtmlParsing(bXhtmlParsing)
end

function EditText:SetXhtmlText(sText)

	self.m_udNativeInstance:SetXhtmlText(sText)
end

--- <summary>
--- Optional - used if StartEditing() without complete arguments is called.
--- </summary>


--- <summary>
--- Optional - used if StartEditing() without complete arguments is called.
--- </summary>


--- <summary>
--- Optional - used if StartEditing() without complete arguments is called.
--- </summary>


function EditText:StartEditing(
	eventReceiver,
	bAllowNonLatinKeyboard)

	local sDescription = self.EditingDescription or ''--[[String.Empty]]
	local iMaxCharacters = self.EditingMaxChars or 0
	local sRestrict = self.EditingRestrictedCharacters or ''--[[String.Empty]]

	return self:StartEditing5_80E305EB(eventReceiver, sDescription, iMaxCharacters, sRestrict, bAllowNonLatinKeyboard)
end

function EditText:StartEditing5_80E305EB(
	eventReceiver,
	sDescription,
	iMaxCharacters,
	sRestrict,
	bAllowNonLatinKeyboard)

	local bReturn = self.m_udNativeInstance:StartEditing(
		eventReceiver:get_NativeInstance(),
		sDescription,
		iMaxCharacters,
		sRestrict,
		bAllowNonLatinKeyboard)

	-- Focus in on first start/successful start.
	if bReturn then

		self:DispatchEvent(self, 'focusIn'--[[FocusEvent.FOCUS_IN]])
	end

	return bReturn
end

function EditText:StopEditing()

	self.m_udNativeInstance:StopEditing()
end
return EditText
